Fix O2M update permission checking

hanneskuettner opened the pull request:
<!--

Heya! Thanks for opening a Pull Request! If your PR is implementing a new feature or fix for Directus, please make sure your PR adheres to the following requirements:

- The PR closes an Issue (not Discussion)
- Tests are added/updated and are passing locally if applicable
- Documentation was added/updated if applicable

Please make sure to "Link" the issue you're closing. Without a Linked issue, this PR won't be accepted. See https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue for more information.

-->

## Scope

This PR tackles two problems: 

* O2M fields were not included in the fetching of permission flags, resulting in a permission error whenever a user tried to perform an update operation on a O2M related field of a parent collection item
* O2M updates are performed using the `updateMany` method of the `ItemsService` and the payload contains the primary key of the related item that is being updated. Due to the simple field extraction logic for the fields that are passed to `validateAccess` this would result in a permissions error if the user does not have update permissions for the primary key of the related collection. 

What's changed:

- Replace `getAstFromQuery` with a simplified AST construction where every field (no matter if it is a related field or not) is treated as "simple" `FieldNode`. This, together with `fetchPermittedAstRootFields` fetches the permission flags for each individual field as needed for checking field level permissions for update and create actions.
- Prevent passing in the primary key of a related item when updating it through a parent item relation, as the primary key can never change  (https://github.com/directus/directus/pull/23915/commits/97055314da0510052ac1966fad81bd08ad3d3be4)

## Potential Risks / Drawbacks

- ?

## Review Notes / Questions

- ~~The inclusion of the `o2mNodes` isn't strictly necessary after a rework of the PR, which only includes `FieldNode` nodes in the AST, but doesn't hurt either.~~

hanneskuettner committed:
Fetch permissions corresponding to the validated action

hanneskuettner committed:
Fetch permission flags for O2M relations

hanneskuettner committed:
Only validate access against the primary key field if we can reasonably assume that the user want's to update it

hanneskuettner committed:
Filter out `null` selects

hanneskuettner committed:
Fix unit test

paescuj committed:
Respect action for fields check

Before this, "read" permissions were used when checking permissions for
"updates".

paescuj committed:
Adjust options for parseFields in import-export.ts

hanneskuettner committed:
Fix comment typo

br41nslug reviewed (dismissed):
Nothing to comment on the code 🚀 and fixes the issue in my tests (tested on postgres, sqlite)

paescuj committed:
Properly fix issue with update requiring pk perms

While in commit 918de2f we ensured the correct perms ("update" perms
instead of "read" perms) are checked for rel item updates, now the update
permission on the primary key field was still required since the primary
key has been part of the payload. This then failed a bit later in the
payload field check performed in the `processPayload` function.
So the proper fix here is to omit the primary key from the payload for
updates (primary key cannot change anyway). With this we can also drop
the fields filter workaround in the `updateMany` method.

Tested with all relationship types.

hanneskuettner committed:
Construct the AST ourselves, as we are only interested in all the top level fields of an item, regardless if they relate to any actual DB backed field

hanneskuettner committed:
Fix test

paescuj committed:
Try to fix import issue in tests

paescuj committed:
Merge branch 'main' into hannes/ser-364

paescuj committed:
Adjust recently added test cases

paescuj committed:
Clean-up

Basically reverting changes from 918de2f (respecting action) as they are
no longer required with 2290444 (constructing own AST for validating
item access)

paescuj committed:
Inline children to make AST construction clearer

hanneskuettner committed:
Remove o2mNodes from fetchPermittedAstRootFields as it is no longer needed

hanneskuettner committed:
Remove unused import

hanneskuettner committed:
Revert changes in `getDBQuery` since they aren't strictly necessary anymore, since every node in the AST is a `FieldNode` now

br41nslug reviewed (approved):
code change looks good and couldnt find an case breaking in my tests so far 😄
